cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -1066,20 +1098,91 @@ class ObjectAppendStream final : public io::OutputStream {
       // flush. This also avoids some unhandled errors when flushing in the destructor.
       return Status::OK();
     }
-    return CommitBlockList(block_blob_client_, block_ids_, commit_block_list_options_);
+
+    auto fut = FlushAsync();
+    RETURN_NOT_OK(fut.status());
+
+    return CommitBlockList();
```

Comment:
I think it's better to keep the `CommitBlockList` version that takes parameters and have all callers acquire the lock before calling it.

Reply:
In general, keep lock acquisition "in the outside" as much as possible.

Reply:
Addressed your feedback in [4ce4ea1](https://github.com/apache/arrow/pull/43096/commits/4ce4ea1520c90ed70b530d88b9042e1a468fa77f)